File: framework/service/bridge_error.go
package gnonative

import (
	"encoding/json"
	stderrors "errors"

	"connectrpc.com/connect"
	"github.com/pkg/errors"

	api_gen "github.com/gnolang/gnonative/v4/api/gen/go"
)

// BridgeErrorPrefix marks a rejection message that carries a JSON envelope
// rather than plain text.
//
// The mobile bridge can only hand a string to the host platform ("Only the
// reject()'s message argument will be thrown to React Native"), so a connect
// error's details cannot cross as structured data, and the ErrCode would survive
// only as the "ErrOutOfGas(#211)" shape ErrCode.Error() renders. The envelope
// carries the detail instead, and this marks it, so a reader tells one from an
// ordinary message without parsing every rejection as JSON.
//
// No version: the Go framework and the TypeScript package that reads it ship in
// the same npm release, so the two cannot disagree about the format.
const BridgeErrorPrefix = "gnonative-error:"

// bridgeErrorEnvelope is the payload following BridgeErrorPrefix. It mirrors the
// published ErrDetails, so a reader decodes it into its own generated types, and
// keeps the underlying error text alongside for logs.
type bridgeErrorEnvelope struct {
	Detail bridgeErrorDetail `json:"detail"`
	Error  string            `json:"error"`
}

type bridgeErrorDetail struct {
	Code    api_gen.ErrCode `json:"code"`
	Message string          `json:"message"`
}

// bridgeError renders err for the bridge, preserving its ErrDetails when it has
// one. `context` names which bridge call failed, as the plain wrapping did.
//
// Every path that rejects a promise goes through this, not only the unary one: a
// streaming call carries the result of a broadcast.
//
// An error without a code is returned as before — an envelope naming no code
// would cost the plain message and add nothing.
func bridgeError(err error, context string) error {
	wrapped := errors.Wrap(err, context)

	detail := errDetailOf(err)
	if detail == nil {
		return wrapped
	}

	payload, marshalErr := json.Marshal(bridgeErrorEnvelope{
		Detail: bridgeErrorDetail{Code: detail.Code, Message: detail.Message},
		Error:  wrapped.Error(),
	})
	if marshalErr != nil {
		// Losing the detail degrades the message; dropping the error would hide it.
		return wrapped
	}

	return errors.New(BridgeErrorPrefix + string(payload))
}

// errDetailOf reads the ErrDetails a connect error carries, or nil. It has to
// come from the details: the error crossed connect from the in-process server,
// so the typed chain api_gen.Codes walks is gone by now.
func errDetailOf(err error) *api_gen.ErrDetails {
	var connectErr *connect.Error
	if !stderrors.As(err, &connectErr) {
		return nil
	}

	for _, detail := range connectErr.Details() {
		value, valueErr := detail.Value()
		if valueErr != nil {
			continue
		}
		if details, ok := value.(*api_gen.ErrDetails); ok {
			return details
		}
	}

	return nil
}

File: examples/js/react-native/hello/grpc/bridge_error.ts
import {Code, ConnectError} from '@connectrpc/connect';

/**
 * Marker for a bridge rejection carrying a JSON envelope rather than plain text.
 * Must match BridgeErrorPrefix in framework/service/bridge_error.go.
 */
const BRIDGE_ERROR_PREFIX = 'gnonative-error:';

type BridgeErrorEnvelope = {
  detail?: {code?: number; message?: string};
  error?: string;
  connectCode?: number;
};

/**
 * Unwraps a bridge rejection into a ConnectError, or returns it untouched.
 *
 * This example pins a generated API from before ErrDetails carried a code, so it
 * takes only the message out of the envelope rather than rebuilding the typed
 * detail. Use @gnolang/gnonative's errDetailOf if you want the code itself.
 */
export function bridgeErrorToConnectError(error: unknown): unknown {
  const message = messageOf(error);
  if (!message?.startsWith(BRIDGE_ERROR_PREFIX)) return error;

  let envelope: BridgeErrorEnvelope;
  try {
    envelope = JSON.parse(
      message.slice(BRIDGE_ERROR_PREFIX.length),
    ) as BridgeErrorEnvelope;
  } catch {
    // A malformed envelope must not lose the failure it describes.
    return error;
  }

  const code = envelope.connectCode;
  return new ConnectError(
    envelope.error ?? envelope.detail?.message ?? message,
    typeof code === 'number' && code in Code ? code : Code.Unknown,
    undefined,
    undefined,
    error,
  );
}

function messageOf(error: unknown): string | undefined {
  if (typeof error === 'string') return error;
  if (error instanceof Error) return error.message;
  if (error && typeof error === 'object') {
    const message = (error as {message?: unknown}).message;
    if (typeof message === 'string') return message;
  }
  return undefined;
}
